<a href="https://colab.research.google.com/github/Vedant-cloud-ctrl/Quantium/blob/main/Quantium_Task1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
upload = files.upload()

Saving QVI_transaction_data(1).csv to QVI_transaction_data(1).csv
Saving QVI_purchase_behaviour.csv to QVI_purchase_behaviour.csv


In [2]:
import pandas as pd
import numpy as np
import io

In [3]:
df_1 = pd.read_csv(io.BytesIO(upload["QVI_transaction_data(1).csv"]))
df_2 = pd.read_csv(io.BytesIO(upload["QVI_purchase_behaviour.csv"]))

In [31]:
df_1.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2019-05-20          1            1343     383        61   
3 2018-08-17          2            2373     974        69   
4 2018-08-18          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

In [5]:
df_2.head()

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream

In [6]:
df_1.dtypes

DATE                int64
STORE_NBR           int64
LYLTY_CARD_NBR      int64
TXN_ID              int64
PROD_NBR            int64
PROD_NAME          object
PROD_QTY            int64
TOT_SALES         float64
dtype: object

The DATE column type is integer we'll convert into datetime format before proceeding further.

In [7]:
# Convert DATE column in datetime format
df_1["DATE"] = pd.to_datetime(df_1["DATE"], origin="1899-12-30", unit="D")

We need check if the data contains right products.

In [11]:
df_1["PROD_NAME"]

0           Natural Chip        Compny SeaSalt175g
1                         CCs Nacho Cheese    175g
2           Smiths Crinkle Cut  Chips Chicken 170g
3           Smiths Chip Thinly  S/Cream&Onion 175g
4         Kettle Tortilla ChpsHny&Jlpno Chili 150g
                            ...                   
264831     Kettle Sweet Chilli And Sour Cream 175g
264832               Tostitos Splash Of  Lime 175g
264833                    Doritos Mexicana    170g
264834     Doritos Corn Chip Mexican Jalapeno 150g
264835               Tostitos Splash Of  Lime 175g
Name: PROD_NAME, Length: 264836, dtype: object

**Text Analysis**

The product contains all potato chips lets analyze it further to check if  the data has more than a chips by using basic text analysis

In [16]:
# Tokenize given words for product name
product_words = df_1["PROD_NAME"].str.split().explode().str.lower()
print("Tokenized words of poduct name:")
product_words

Tokenized words of poduct name:


0             natural
0                chip
0              compny
0         seasalt175g
1                 ccs
             ...     
264835       tostitos
264835         splash
264835             of
264835           lime
264835           175g
Name: PROD_NAME, Length: 1359084, dtype: object

In [18]:
from collections import Counter
# Count word frequencies
word_feq = Counter(product_words)
common_words = word_feq.most_common(30)
print("Common words in product name:")
for word, freq in common_words:
  print(f"{word}: {freq}")

Common words in product name:
175g: 60561
chips: 49770
150g: 43131
kettle: 41288
&: 35565
smiths: 28860
salt: 27976
cheese: 27890
pringles: 25102
134g: 25102
doritos: 24962
crinkle: 23960
110g: 22387
corn: 22063
original: 21560
cut: 20754
chip: 18645
170g: 18502
salsa: 18094
chicken: 15407
165g: 15297
sea: 14145
thins: 14075
sour: 13882
crisps: 12607
330g: 12540
vinegar: 12402
chilli: 12389
300g: 12041
rrd: 11894


In [20]:
# Remove words with symbols and columns
cleaned_words = product_words[product_words.str.contains(r"^[a-zA-Z]+$", na=False)]
cleaned_words_counts = Counter(cleaned_words).most_common(30)
print("Cleaned words from product name:")
for word, counts in cleaned_words_counts:
  print(f"{word}: {counts}")

Cleaned words from product name:
chips: 49770
kettle: 41288
smiths: 28860
salt: 27976
cheese: 27890
pringles: 25102
doritos: 24962
crinkle: 23960
corn: 22063
original: 21560
cut: 20754
chip: 18645
salsa: 18094
chicken: 15407
sea: 14145
thins: 14075
sour: 13882
crisps: 12607
vinegar: 12402
chilli: 12389
rrd: 11894
infuzions: 11057
supreme: 10963
ww: 10320
cobs: 9693
popd: 9693
tortilla: 9580
tostitos: 9471
twisties: 9454
sensations: 9429


From the given text analysis there are salsa product we'er interested in chips category, so lets remove it.

In [32]:
# Remove salsa products
df_1 = df_1[~df_1.PROD_NAME.str.lower().str.contains("salsa", na=False)]
df_1["PROD_NAME"]

0           Natural Chip        Compny SeaSalt175g
1                         CCs Nacho Cheese    175g
2           Smiths Crinkle Cut  Chips Chicken 170g
3           Smiths Chip Thinly  S/Cream&Onion 175g
4         Kettle Tortilla ChpsHny&Jlpno Chili 150g
                            ...                   
264831     Kettle Sweet Chilli And Sour Cream 175g
264832               Tostitos Splash Of  Lime 175g
264833                    Doritos Mexicana    170g
264834     Doritos Corn Chip Mexican Jalapeno 150g
264835               Tostitos Splash Of  Lime 175g
Name: PROD_NAME, Length: 246742, dtype: object

We'll summarize the data to find any null values or any outliers

In [41]:
df_1.describe()

DATE      STORE_NBR  LYLTY_CARD_NBR  \
count                         246742  246742.000000    2.467420e+05   
mean   2018-12-30 01:19:01.211467520     135.051098    1.355310e+05   
min              2018-07-01 00:00:00       1.000000    1.000000e+03   
25%              2018-09-30 00:00:00      70.000000    7.001500e+04   
50%              2018-12-30 00:00:00     130.000000    1.303670e+05   
75%              2019-03-31 00:00:00     203.000000    2.030840e+05   
max              2019-06-30 00:00:00     272.000000    2.373711e+06   
std                              NaN      76.787096    8.071528e+04   

             TXN_ID       PROD_NBR       PROD_QTY      TOT_SALES  
count  2.467420e+05  246742.000000  246742.000000  246742.000000  
mean   1.351311e+05      56.351789       1.908062       7.321322  
min    1.000000e+00       1.000000       1.000000       1.700000  
25%    6.756925e+04      26.000000       2.000000       5.800000  
50%    1.351830e+05      53.000000       2.000000       7.400000  
75%    2.026538e+05      87.000000       2.000000       8.800000  
max    2.415841e+06     114.000000     200.000000     650.000000  
std    7.814772e+04      33.695428       0.659831       3.077828

In [43]:
df_1.min()

DATE              2018-07-01 00:00:00
STORE_NBR                           1
LYLTY_CARD_NBR                   1000
TXN_ID                              1
PROD_NBR                            1
PROD_NAME           Burger Rings 220g
PROD_QTY                            1
TOT_SALES                         1.7
dtype: object

In [44]:
df_1.max()

DATE                         2019-06-30 00:00:00
STORE_NBR                                    272
LYLTY_CARD_NBR                           2373711
TXN_ID                                   2415841
PROD_NBR                                     114
PROD_NAME         Woolworths Cheese   Rings 190g
PROD_QTY                                     200
TOT_SALES                                  650.0
dtype: object

In [49]:
df_1["PROD_QTY"].describe()

count    246742.000000
mean          1.908062
std           0.659831
min           1.000000
25%           2.000000
50%           2.000000
75%           2.000000
max         200.000000
Name: PROD_QTY, dtype: float64

There no null values in the data but the product quantity has oultier lets investigate further.

In [52]:
df_1[df_1["PROD_QTY"] == 200]

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
69762 2018-08-19        226          226000  226201         4   
69763 2019-05-20        226          226000  226210         4   

                              PROD_NAME  PROD_QTY  TOT_SALES  
69762  Dorito Corn Chp     Supreme 380g       200      650.0  
69763  Dorito Corn Chp     Supreme 380g       200      650.0

We found out that the product quantity has been purchased by same customer, let's find out if the same customer has purchased any other products

In [53]:
df_1[df_1["LYLTY_CARD_NBR"] == 226000]

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
69762 2018-08-19        226          226000  226201         4   
69763 2019-05-20        226          226000  226210         4   

                              PROD_NAME  PROD_QTY  TOT_SALES  
69762  Dorito Corn Chp     Supreme 380g       200      650.0  
69763  Dorito Corn Chp     Supreme 380g       200      650.0

looks like the customer had only two transaction over a year assuming that it is not an ordinary retail customer might be buying chips for commercial purpose. We'll drop the customer

In [54]:
# Remove customer
df_1 = df_1[df_1["LYLTY_CARD_NBR"] != 226000]

In [63]:
if 226000 in df_1["LYLTY_CARD_NBR"].values:
  print("Customer still exists")
else:
  print("Customer not found")

Customer not found


In [64]:
df_1.max()

DATE                         2019-06-30 00:00:00
STORE_NBR                                    272
LYLTY_CARD_NBR                           2373711
TXN_ID                                   2415841
PROD_NBR                                     114
PROD_NAME         Woolworths Cheese   Rings 190g
PROD_QTY                                       5
TOT_SALES                                   29.5
dtype: object

Now let see the transaction order over time to see if there is any data issue or missing data.

In [67]:
df_1.groupby("DATE").size().reset_index(name="N")

DATE    N
0   2018-07-01  663
1   2018-07-02  650
2   2018-07-03  674
3   2018-07-04  669
4   2018-07-05  660
..         ...  ...
359 2019-06-26  657
360 2019-06-27  669
361 2019-06-28  673
362 2019-06-29  703
363 2019-06-30  704

[364 rows x 2 columns]

Its 364 rows, means that the data is till 364 date which indicates one missing value from 1 June 2018 to 30 July 2019. Lets visulaize it to understand each transaction and missing data.

In [69]:
import matplotlib.pyplot as plt
import seaborn as sns